src/arkode/arkode_sprkstep.c

```diff
@@ -0,0 +1,781 @@
+                    MSG_ARK_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  ark_mem = (ARKodeMem)arkode_mem;
+  return (arkRootInit(ark_mem, nrtfn, g));
+}
+
+/* Utility to call f1 and increment the counter */
+int sprkStep_f1(ARKodeSPRKStepMem step_mem, sunrealtype tcur, N_Vector ycur,
+                N_Vector f1, void* user_data)
+{
+  int retval = step_mem->f1(tcur, ycur, f1, user_data);
+  step_mem->nf1++;
+  return retval;
+}
+
+/* Utility to call f2 and increment the counter */
+int sprkStep_f2(ARKodeSPRKStepMem step_mem, sunrealtype tcur, N_Vector ycur,
+                N_Vector f2, void* user_data)
+{
+  int retval = step_mem->f2(tcur, ycur, f2, user_data);
+  step_mem->nf2++;
+  return retval;
+}
+
+/*---------------------------------------------------------------
+  sprkStep_FullRHS:
+
+  This is just a wrapper to call the user-supplied RHS,
+  f1(t,y) + f2(t,y).
+
+  This will be called in one of three 'modes':
+    ARK_FULLRHS_START -> called at the beginning of a simulation
+                         or after post processing at step
+    ARK_FULLRHS_END   -> called at the end of a successful step
+    ARK_FULLRHS_OTHER -> called elsewhere (e.g. for dense output)
+
+  If it is called in ARK_FULLRHS_START mode, we store the vectors
+  f1(t,y) and f2(t,y) in sdata for possible reuse in the first stage
+  of the subsequent time step.
+
+  If it is called in ARK_FULLRHS_END mode and the method coefficients
+  support it, we may just copy the stage vectors to fill f instead
+  of calling f().
+
+  ARK_FULLRHS_OTHER mode is only called for dense output in-between
+  steps, so we strive to store the intermediate parts so that they
+  do not interfere with the other two modes.
+  ---------------------------------------------------------------*/
+int sprkStep_FullRHS(void* arkode_mem, realtype t, N_Vector y, N_Vector f,
+                     int mode)
+{
+  int retval                 = 0;
+  ARKodeMem ark_mem          = NULL;
+  ARKodeSPRKStepMem step_mem = NULL;
+
+  /* access ARKodeSPRKStepMem structure */
+  retval = sprkStep_AccessStepMem(arkode_mem, "SPRKStep_FullRHS", &ark_mem,
+                                  &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* perform RHS functions contingent on 'mode' argument */
+  switch (mode)
+  {
+  case ARK_FULLRHS_START:
+  case ARK_FULLRHS_END:
+  case ARK_FULLRHS_OTHER:
+
+    retval = sprkStep_f1(step_mem, t, y, f, ark_mem->user_data);
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                      "SPRKStep_FullRHS", MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
+    retval = sprkStep_f2(step_mem, t, y, f, ark_mem->user_data);
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                      "SPRKStep_FullRHS", MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
```

Comment:
This only works if the problem is appropriately partitioned.  I don't recall the documentation giving any stern warnings against users attempting to provide non-partitioned RHS functions.

Reply:
I added warnings.